### PR TITLE
Claude 3

### PR DESCRIPTION
**Pull request by cunnymessiah**
Implements support for Claude 3 models and the new messaging API.

Will add the vision functionality in a bit.

### CODE REVIEW

**Comment by Cohee1207** [comment]
"name" property should be removed from converted messages if "Add character names" is enabled. Claude doesn't like that.

**Comment by Cohee1207** [comment]
Presumably, Claude also started rejecting legacy aliases for older models. Should we fix it here?
```
model: "claude-v1.3" is a legacy model alias and is not supported in this API. Did you mean "claude-1.3"?
```

**Comment by cunnymessiah** [comment]
If we want to split up implementations then sure, 1.3 can be routed back to the completions api while all 2.0+ requests get routed to messages. if 1.3 still works on the old api (not sure haven't used 1.3 in ages)

**Comment by cunnymessiah** [comment]
> "name" property should be removed from converted messages if "Add character names" is enabled. Claude doesn't like that.

should i just add it to the prompt then instead? i.e char: message

 not really familiar with what this option did since i never used it

**Comment by Cohee1207** [comment]
> should i just add it to the prompt then instead? i.e char: message

This is how the older prompt converter used to work. So yes, I guess.

**Comment by Cohee1207** [comment]
> If we want to split up implementations then sure, 1.3 can be routed back to the completions api while all 2.0+ requests get routed to messages. if 1.3 still works on the old api (not sure haven't used 1.3 in ages)

I just checked that it works with messages endpoint for as far as claude-instant-1.0, it only requires applying a new model naming scheme to use them.

**Comment by cunnymessiah** [comment]
> it only requires applying a new model naming scheme to use them.

wdym by this?

**Comment by Cohee1207** [comment]
The error message similar to what I provided above usually tells what model name should be specified. For most cases, it could be fixed by removing the `v`.

`claude-v1.3` -> `claude-1.3`

My best guess is that going through all versions listed in the dropdown one by one and fixing them would be a solution since I haven't found a full list of legacy models in the API docs.

**Comment by cunnymessiah** [comment]
oh that's good then, we can just interact with one endpoint. i should have read the error message 🤦‍♂️

doesn't seem like 1.2 works though, same story for 1.0 instant. should we just remove them?

also seems like 1.3-100k doesn't work, ill test if 1.3 is routing to the 100k version by default or just the base 9k

**Comment by BasadoBased** [comment]
1.3 and earlier were supposed to be retired completely on feb 29 2024 (5 days ago)

**Comment by cunnymessiah** [comment]
> 1.3 and earlier were supposed to be retired completely on feb 29 2024 (5 days ago)

anthropic commonly says one thing and does the opposite

@Cohee1207 since there isn't a 'latest' opus or sonet endpoint i just added the snapshot in the latest tab instead. i imagine they'd make one eventually right?

**Comment by Cohee1207** [comment]
So I applied the empirical rule and just removed v from all model IDs.

image

Here's what doesn't work from this list:
- claude-1.0
- claude-instant-1.1-100k
- claude-1.3-100k
- claude-instant-1-100k
- claude-instant-1
- claude-1-100k
- claude-1
- claude-2

Basically all `100k`s and non-snapshot aliases of v2 and older. Everything else still works, including all 9k instant snapshots. Could be on my key though and for some it's already gone, hard to tell.

UPD: Fixed screenshot.

**Comment by cunnymessiah** [comment]
alright then that should be the text part done, i'll do vision now

**Comment by Yardanico** [comment]
I wanted to add one important thing - the official Claude API doesn't actually have non-snapshotted versions like "claude-3-opus"
```
anthropic.NotFoundError: Error code: 404 - {'type': 'error', 'error': {'type': 'not_found_error', 'message': 'model: claude-3-opus'}}
```

**Comment by cunnymessiah** [comment]
should be finished @Cohee1207 give it a try. i've played around with the vision stuff for about half an hour and it seems all good.

File: src/endpoints/anthropic.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+const express = require('express');
+const { jsonParser } = require('../express-common');
+
+const router = express.Router();
+
+router.post('/caption-image', jsonParser, async (request, response) => {
+    try {
+        const mimeType = request.body.image.split(';')[0].split(':')[1];
+        const base64Data = request.body.image.split(',')[1];
+        const url = 'https://api.anthropic.com/v1/messages';
```
Comment (Cohee1207) [inline src/endpoints/anthropic.js]: Is the reverse proxy setting ignored here? I saw the frontend part check, but here it's always Anthropic

**Comment by Cohee1207** [comment]
Just to confirm.  Am I right that the "Exclude Human/Assistant prefixes" and "Exclude Assistant suffix" settings are now unused with the messages API and could be hidden/removed? They seem very specific to manual prompt formatting, which won't be available.

**Comment by Cohee1207** [comment]
1. Vision - works.
2. Prefill - works.
3. All listed models - work.

Only left to cover the last 2 questions and we're good to go and merge.

**Comment by DonMoralez** [comment]
> Just to confirm. Am I right that the "Exclude Human/Assistant prefixes" and "Exclude Assistant suffix" settings are now unused with the messages API and could be hidden/removed? They seem very specific to manual prompt formatting, which won't be available.

"Exclude Human/Assistant prefixes" - unusable with the messages API, it is easier to remove it. 
"Exclude Assistant suffix" - used when you want to manually create a prefill (or several prefills for quick switching or testing) directly in the prompt manager, so can still be used.

**Comment by cunnymessiah** [comment]
> Just to confirm. Am I right that the "Exclude Human/Assistant prefixes" and "Exclude Assistant suffix" settings are now unused with the messages API and could be hidden/removed? They seem very specific to manual prompt formatting, which won't be available.

i don't mind either way, the first option i've never used even on the old api so i'm not sure what the point of it is/was and wouldn't mind it being removed. the second option i've only ever used in the past.

> Is the reverse proxy setting ignored here? I saw the frontend part check, but here it's always Anthropic

correct this should be changed to whatever the reverseproxy url is, vision isn't supported in khanon's proxy build just yet but he plans to implement that sometime soon.

> "Exclude Assistant suffix" - used when you want to manually create a prefill (or several prefills for quick switching or testing) directly in the prompt manager, so can still be used.

fair enough, i'll leave it in then, although it'll only be effective just on the frontend side of things since it only serves to hide the prefill block with how the messages api is setup (there is no more assistant postfix).

i'll make those changes and then we'll merge

**Comment by DonMoralez** [comment]
> fair enough, i'll leave it in then, although it'll only be effective just on the frontend side of things since it only serves to hide the prefill block with how the messages api is setup (there is no more assistant postfix). 

What I mean is you can rename it to "Use prefill" adding the corresponding functionality.  Without this checkbox, you will have to clear the prefill text box every time you want to add a prefill manually using the prompt manager.
P.S. In theory, you can remove both the checkbox and the prefill text box, to simplify tavern. But from what I've been seeing on 2ch, there will be a ton of cursing, since it's easier to blame the tavern than it is to find out how everything works.

**Comment by Cohee1207** [comment]
I'd vote for a mix of both: remove the checkbox, but leave the prefill box.
1. "Exclude Assistant suffix" doesn't make sense anymore in a way it is named and described.
2. As for the prefill, a lot of users are looking for it, like they hear it from someone's guide but don't know exactly where to look at, as it doesn't explain it.

**Comment by cunnymessiah** [comment]
> "Exclude Assistant suffix" doesn't make sense anymore in a way it is named and described.

agreed.

> What I mean is you can rename it to "Use prefill" adding the corresponding functionality.

i like this idea, that's pretty much what the option does now anyway, it has no bearing on the backend anymore other than stopping the prefill from being sent

**Comment by Cohee1207** [comment]
Done. Let it slide.

**Comment by DonMoralez** [comment]
> 2. As for the prefill, a lot of users are looking for it, like they hear it from someone's guide but don't know exactly where to look at, as it doesn't explain it.

There are special boards where people discuss things like this. Very often, they make false conclusions there, and, well, it leads to such things.

**Comment by cunnymessiah** [comment]
> > 2. As for the prefill, a lot of users are looking for it, like they hear it from someone's guide but don't know exactly where to look at, as it doesn't explain it.
> 
> There are special boards where people discuss things like this. Very often, they make false conclusions there, and, well, it leads to such things.

im sorry bro... i left the option in but cohee blasted it from outter space. lel

**Comment by Cohee1207** [comment]
What the option in the end did was literally just hide the prefill box and disable it from being sent. It is equivalent to removing everything from the prefill box, but more convoluted.

**Comment by DonMoralez** [comment]
By the way, I'm not quite sure, but it seems that the system prompt behaves quite strangely, at least on AWS. I'll conduct more tests a bit later, but it appears that the AI either ignores or partly ignores information from the system prompt, but works just fine without it.

**Pull request merged**